postconfirmc.1
." Text automatically generated by txt2man-1.4.7
.TH untitled  "July 10, 2017" "" ""
.SH NAME
\fBpostconfirm \fP- Mail confirmation system
\fB
.SH SYNOPSIS
.nf
.fam C
\fBpostconfirmc\fP [\fB-h\fP] [\fB--stop\fP] [\fB--socket\fP \fISOCKET\fP]
[{ \fIconfirm\fP | \fIdmarc-rewrite\fP | \fIdmarc-reverse\fP }]
.PP
\fBpostconfirmd\fP [\fIoptions\fP]
.fam T
.fi
.SH DESCRIPTION

postconfirm consists of \fBpostconfirmd\fP, which is the long-running
(daemon) part, and \fBpostconfirmc\fP, which is the client part of a
client-server program which handles email confirmations. It is
intended as a front-end to mailing lists. It provides
funcitonality which is a subset of TMDA, but is adapted to
high-volume usage and does not have anywhere near all the bells
and whistles which TMDA has. On the other hand, since the
allowlist lookup is done by the long-running server part, the
overhead of doing a verification that a poster has a confirmed
address is much smaller than for TMDA.
.PP
Early tests indicate that with a allowlist of 16000 entries a lookup
costs about 0.01 seconds and about ~5 Mbytes for the server daemon, in
contrast whith TMDA which cost between ~10s and ~600s and ~98 Mb *per
lookup* as it was set up in front of the IETF lists.
.SH ACTIONS
\fIconfirm\fP (the default action):
.PP
Somewhat simplified, the flow for regular mail is as follows:
.PP
.nf
.fam C
     mail agent
         |
         v
     postconfirm/mailman wrapper [copy] -> cached in cache dir
         |
         v
     postconfirmc <==> postconfirmd
         |
         v
     postconfirm/mailman wrapper [allow] -> Mailman
       [grey]
         |
         v
     (waiting for confirmation)

.fam T
.fi
In the start of the flow above, a message comes in. If the envelope
From is in the lists known by \fBpostconfirmd\fP, it is passed through the
mailman/ postconfirm wrapper, where \fBpostconfirmc\fP is called to talk
with \fBpostconfirmd\fP and get either an OK to pass the message to Mailman,
or cache it.
.PP
In more detail, this is the sequence of actions that postconfirm goes
through when processing an incoming email message:
.IP 1. 4
The mail is put in the cache directory.
.IP 2. 4
If the message is a delivery-status report (content-type
multipart/report, report-type delivery-status) then this is logged,
and no further action is taken.
.IP 3. 4
If the precedence matches the bulk_regex setting in the
configuration file, then:
.RS
.PP
a. if the envelope sender is in the allowlist or the allow_regex
list, then the message is forwarded directly.
.PP
b. if not, the receipt of a bulk message is logged, and no further
action is taken.
.RE
.IP 4. 4
If the message has an auto-submitted header, and the value matches
the auto_submitted_regex in the configuration file, then:
.RS
.PP
a. if the envelope sender is in the allowlist or the allow_regex
list, then the message is forwarded directly.
.PP
b. if not, the receipt of a bulk message is logged, and no further
action is taken.
.RE
.IP 5. 4
If the subject matches the regex for postconfirm's confirmation
email subject, the confirmation is processed, and then:
.RS
.PP
a. if the confirmation is valid, the held message is forwarded,
and the confirmed address is added to the allowlist.
.PP
b. if not, the failed confirmation is logged, and the message
is saved.
.RE
.IP 6. 4
If the envelope sender is in the allowlist or the allow_regex list,
the message is forwarded.
.IP 7. 4
If there was no match earlier, a confirmation request is sent out
for the message.
.PP
\fIdmarc-rewrite\fP:
.PP
For use with mailing lists, aliases, etc., where the presence of
strict dmarc rules will prevent acceptance of the list messages
unless rewriting is done.
.PP
\fIdmarc-reverse\fP:
.PP
The inverse operation of \fIdmarc-rewrite\fP. Undoes the rewrite, 
and replaces the rewritten address with the original address.
.SH OPTIONS

{ \fIconfirm\fP | \fIdmarc-rewrite\fP | \fIdmarc-reverse\fP }
the action to take
.TP
.B
\fB-h\fP, \fB--help\fP
show this help message and exit
.TP
.B
\fB--stop\fP
stop the \fBpostconfirmd\fP daemon
.TP
.B
\fB--socket\fP \fISOCKET\fP
use the given \fISOCKET\fP for client-daemon communication
.SH FILES
\fBpostconfirmd\fP reads its configuration from the following files (if
found), in the given order:
.PP
.nf
.fam C
      <bin-dir>/postconfirm.conf
      /etc/postconfirm.conf
      /etc/postconfirm/postconfirm.conf
      ~/.postconfirmrc

.fam T
.fi
where <bin-dir> is the directory where \fBpostconfirmd\fP is installed.
.SH AUTHOR
Written by Henrik Levkowetz, <henrik@merlot.tools.ietf.org>. Uses the
daemonize module from Chad J. Schroeder, from the Python Cookbook at
http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/278731, and
the readyexec module from Frank J. Tobin, available at
http://readyexec.sourceforge.net/.
.SH COPYRIGHT
Copyright 2008-2017 Henrik Levkowetz
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version. There is NO WARRANTY; not even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE. See the GNU General Public License for more details.


changelog
postconfirm (0.42.1) ietf; urgency=low

  * Added domain name normalization (to lowercase) before checking whether we
    got a hit on the _dmarc domain lookup.  Commented out verbose wall-time
    log messages.

 -- Henrik Levkowetz <henrik@levkowetz.com>  31 Jul 2018 08:48:14 -0700

postconfirm (0.42.0) ietf; urgency=high

  * Added a check on the sender address during dmarc-reverse processing,
    requiring that the sender is whitelisted or has a confirmed address.

  * Changed the logging of dmarc-reverse processing to use a more compact
    log line format, and commented out some debug logging statements.

  * Added exception catching and loggin for failures to open smtp sessions
    to the configured server and port.

  * Fixed a log writing bug.

 -- Henrik Levkowetz <henrik@levkowetz.com>  09 May 2018 14:04:31 -0700

postconfirm (0.41.5) ietf; urgency=high

  * Fixed dmarc_reverse() to not re-send to other addresses than the 
    indicated recipient.

 -- Henrik Levkowetz <henrik@levkowetz.com>  20 Apr 2018 05:04:44 -0700


postconfirm (0.41.4) ietf; urgency=high

  * Fixed a bug in dmarc_rewrite(), where the return values from
    email.utils.getaddresses() was not handled properly.

  * Tweaked some setup error messages

 -- Henrik Levkowetz <henrik@levkowetz.com>  01 Mar 2018 06:47:58 -0800

postconfirm (0.41.3) ietf; urgency=high

  * Added warnings on failure for some post-install actions.

  * Changed the interface to accept multiple recipients when given on the
    command line (for use with dmarc-rewrite).  Changed dmarc-rewrite to
    accept multiple recipients, and renamed the recipient argument for other
    functions that can accept recipient lists.  

  * Changed the -e argument to log arguments to syslog and not exit.

  * Reverted an earlier change of socket ownership handling.

 -- Henrik Levkowetz <henrik@levkowetz.com>  21 Feb 2018 08:29:30 -0800

postconfirm (0.41.2) ietf; urgency=high

  * Added and --echo argument to echo back sys.argv.  Rewrote the logic to 
    find sender and recipient, in order to not throw out SENDER and RECIPIENT 
    in environment when they are set to nothing, which in some circumstances 
    can be valid.

  * Don't enforce uid and gid  on socket in foreground mode, to make 
    testing easier.

  * Corrected the trove classifier for development status.

 -- Henrik Levkowetz <henrik@levkowetz.com>  16 Feb 2018 04:28:30 -0800

postconfirm (0.41.1) ietf; urgency=high

  * Changed default dmarc-rewrite %-encoding character from '%' to '=', and 
    made it configurable.

 -- Henrik Levkowetz <henrik@levkowetz.com>  15 Feb 2018 09:57:40 -0800

postconfirm (0.41.0)

  This the first release uploaded to pypi.python.org, for more standardised
  installation using pip.  It's still not expected to be ready for beta use,
  but installation on the IETF servers should be easier this way.

  The essensial feature of this release is the ability to do experimental
  dmarc rewrites, both for outgoing mailman list messages, and to handle
  replies coming in to the rewritten addresses.  From the commit log:

  * Added dmarc rewrite functionality to postconfirmd.  Also various 
    accumulated tweaks and changes.

  * Refined setup, in preparation for alpha release to pypi.

  * README updated to indicate that this is an alpha release.

  * Refined mailman interface wrapper, with conditional filter actions.

  * Added some new config defaults.  Fixed an issue with the help display.  
    Updated imports.  Added chown for the socket.

  * Refactored service.py to permit multiple actions.  Added option 
    parsing.  Added bounce handling, but commented out.  Added dmarc handling.  
    Added comprehensive docstring for --help messages.

  * Updated the docstring.  Added default config for dmarc.

  * Tweaked sockserver output to straghten multiple lines into one.

 -- Henrik Levkowetz <henrik@levkowetz.com>  14 Feb 2018 16:01:53 +0000


postconfirm (0.40)

  * Added checking of the mailman list archive flag before adding the
    Archived-At: header field.

  * Added a mailman_dir config option, in order to be able to import
    the python mailman libs even if not on the regular python path.

  * Fixed the -s / --socket option, to make it take an argument as intended.

  * Changed the handling of sys.stdout and sys.stderr when running in foreground
    mode -- they should not go to syslog in that case.

  * Added reading of mailman listinfo for all mailman lists on startup and
    on receipt of SIGHUP.

 -- Henrik Levkowetz <henrik@levkowetz.com>  20 Jan 2017 10:36:45 +0100

postconfirm (0.39)

  * Added support for the Archived-At: header field for the IETF MailArchive
    utility

 -- Henrik Levkowetz <henrik@levkowetz.com>  10 Feb 2014 21:28:27 +0100

postconfirm (0.38)

  * Added BATV (draft-levine-smtp-batv) support.

 -- Henrik Levkowetz <henrik@levkowetz.com>  20 Mar 2010 16:49:50 -0700

postconfirm (0.37)

  * Tweaked the log message for dropped bulk messages

  * Added detection of auto-reply messages which contain a valid
    confirmation code in the body of the message, but not in the
    subject line.

 -- Henrik Levkowetz <henrik@levkowetz.com>  25 Sep 2009 15:37:17 +0200

postconfirm (0.36)

  * Moved the bulk mail detection to after the whitelist check.  Some
    OK mail is also bulk mail.

  * Added logging of py module loading, indicating which file it comes from.

 -- Henrik Levkowetz <henrik@levkowetz.com>  25 Sep 2009 11:47:24 +0200

postconfirm (0.35)

  * Tweaked the wrapper to make the 'filter' action work

  * Added detection of bulk mail and a config file option to set
    a regex for the Precedence: header to detect bulk mail.

 -- Henrik Levkowetz <henrik@levkowetz.com>  24 Sep 2009 17:07:41 +0200

postconfirm (0.33)

  * Added a new action, 'filter', which forwards confirmed email to
    /usr/lib/sendmail instead of to mailman.  For use with filtering
    of alias addresses.


postconfirmc.c
/* $Id: readyexec.c,v 1.7 2002/10/09 20:16:28 ftobin Exp $ */
// Copied with slight modifications from Frank Tobin's 'readyexec.c'

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <unistd.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/uio.h>

#define PROGNAME "postconfirm"
#define DEBUG    0

#define DO_SEND_ARGS     1
#define DO_SEND_ENV      1
#define DO_SEND_FD_NAMES 1

#define debug fprintf
#define SOCK_PATH	"/var/run/postconfirm/socket"
#define SOCK_BUF_LEN	256
#define MY_ARGC 1

void error(const char *msg);
void myclose(FILE *stream);
int  myconnect(const char *path);
void myflush(FILE *stream);
void proto_error(const char *msg);
int  read_exit_code(FILE *stream);
void send_args(FILE *stream, int argc, char **argv);
void usage(void);
void verify_expected(FILE *stream, const char *expected);

void  write_netstring(FILE *stream, const void *ptr, size_t len);
char* read_netstring(FILE *stream, size_t *len_ptr);
void  write_netstring_nulled(FILE *stream, const char *ptr);

void write_netint(FILE *stream, int len);
int read_netint(FILE *stream);

void send_fd(int s, int fd);
void send_named_fd(FILE *stream, const char *name, int fd);

void send_env(FILE *stream);
int count_env(void);

extern char **environ;

FILE *debugf;
int  do_debug;

int main(int argc, char **argv)
{
    int cs; /* control socket */
    short int send_stop;
    char sock_path[SOCK_BUF_LEN];
    int i;
    
    FILE *csfile;
    
    debugf = DEBUG ? stderr : fopen("/dev/null", "w");
    
    // default values
    send_stop = 0;
    strncpy(sock_path, SOCK_PATH, SOCK_BUF_LEN);

    for (i = 1; i < argc; i++) {
	if (strcmp("--stop", argv[i]) == 0) { send_stop = 1; }
	if (strcmp("--socket", argv[i]) == 0) {
	    i++;
	    if (i < argc) {
		strncpy(sock_path, argv[i], SOCK_BUF_LEN);
	    } else {
		error("missing argument to --socket");
	    }
	}
    }
    
    cs = myconnect(sock_path);
    
    csfile = fdopen(cs, "w+");
    if (!csfile) { error("fdopen of socket"); }

    if (send_stop)
    {
	write_netstring_nulled(csfile, "stop");
	return 0;
    }
    else
    {
	write_netstring_nulled(csfile, "conduit");
    }
    
    if (DO_SEND_ARGS)
    {
	debug(debugf, "sending args\n");
	send_args(csfile, argc, argv);
    }
    
    if (DO_SEND_ENV)
    {
	debug(debugf, "sending environment\n");
	send_env(csfile);
    }

    send_named_fd(csfile, "stdin",  STDIN_FILENO);
    send_named_fd(csfile, "stdout", STDOUT_FILENO);
    send_named_fd(csfile, "stderr", STDERR_FILENO);
    
    if (shutdown(cs, SHUT_WR) == -1) { error("shutdown"); }
    
    debug(debugf, "reading exit code\n");
    return read_exit_code(csfile);
}


void myflush(FILE *stream)
{
    if (fflush(stream) != 0)  { error("fflush"); }
}


void myclose(FILE *stream)
{
    if (fclose(stream) != 0)  { error("fclose"); }
}


int myconnect(const char *path)
{
    struct sockaddr_un saddr;
    int s;
    
    strncpy(saddr.sun_path, path, sizeof(saddr.sun_path));
    saddr.sun_family = AF_UNIX;
    
    s = socket(PF_UNIX, SOCK_STREAM, 0);
    if (s == -1) { error("socket"); }
    
    if (connect(s, (struct sockaddr *)&saddr, sizeof(saddr)) == -1)
    {
	error("connect");
    }
    
    return s;
}


void send_args(FILE *stream, int argc, char **argv)
{
    int i;
    
    write_netstring_nulled(stream, "args");
    write_netint(stream, argc - MY_ARGC);
    
    for (i = MY_ARGC; i < argc; i++)
    {
	write_netstring_nulled(stream, argv[i]);
    }
}


void send_named_fd(FILE *stream, const char *name, int fd)
{
    debug(debugf, "sending %s descriptor\n", name);

    if (DO_SEND_FD_NAMES)
    {
	write_netstring_nulled(stream, name);
    }
    
    myflush(stream);
    send_fd(fileno(stream), fd);
}


void send_env(FILE *stream)
{
    int i;
    int count;
    
    write_netstring_nulled(stream, "env");
    count = count_env();
    debug(debugf, "counted %d environment variables\n", count);
    write_netint(stream, count);
    
    for (i = 0; i < count; i++)
    {
	write_netstring_nulled(stream, environ[i]);
    }    
}


int count_env()
{
    int i;
    i = 0;
    
    while (environ[i] != NULL)
    {
	i++;
    }
    
    return i;
}


int read_exit_code(FILE *stream)
{
    int i;
    verify_expected(stream, "exit");
    i = read_netint(stream);
    debug(debugf, "read exit code of %d\n", i);
    return i;
}


void write_netstring_nulled(FILE *stream, const char *ptr)
{
    debug(debugf, "sending string '%s'\n", ptr);
    write_netstring(stream, ptr, strlen(ptr));
}


/* netstrings: http://cr.yp.to/proto/netstrings.txt */
void write_netstring(FILE *stream, const void *ptr, size_t len)
{
    char buff[64] = {0};
    
    snprintf(buff, sizeof(buff), "%s", (const char *)ptr);
    
    if (fprintf(stream, "%ld:", len) == 0) { error("fprintf"); }
    if (fwrite(ptr, 1, len, stream) < 1)   { error("fwrite"); }
    if (fputc(',', stream) == EOF)         { error("fputc"); }
}


void write_netint(FILE *stream, int l)
{
    debug(debugf, "sending netint %d\n", l);
    if (fprintf(stream, "%d,", l) == 0) { error("fprintf of netint"); }
}


int read_netint(FILE *stream)
{
    int i;
    char error_buf[256] = {0};
    char error_buf2[256] = {0};

    if (fscanf(stream, "%d,", &i) < 1)
    {
	fread(error_buf2, 10, 1, stream);
	snprintf(error_buf, sizeof(error_buf),
		 "netint (%s) is invalid", error_buf2);
	proto_error(error_buf);
    }
    
/* debug(debugf, "read a netint of %d\n", i); */
    return i;
}


/* netstrings: http://cr.yp.to/proto/netstrings.txt */
/* bonus: I null-terminate */
char* read_netstring(FILE *stream, size_t *len_ptr)
{
    char error_buf[256] = {0};
    char error_buf2[256] = {0};
    char *buf;
    int c;
    int num_read;
    
    num_read = fscanf(stream, "%4ld:", len_ptr);
    if (num_read < 1)
    {
	if (num_read == EOF) { proto_error("EOF during netstring length"); }
	
	fread(error_buf2, 10, 1, stream);
	snprintf(error_buf, sizeof(error_buf),
		 "netstring length (%s) is invalid",
		 error_buf2);
	proto_error(error_buf);
    }
    
    buf = malloc(*len_ptr + 1); /* malloc(0) is not portable
				   and I want it null terminated */
    
    if (!buf)                                { error("malloc"); }
    bzero(buf, *len_ptr +1);
    if (fread(buf, 1, *len_ptr, stream) < 1) { error("fread"); }
    
    c = fgetc(stream);
    if (c != ',')
    {
	snprintf(error_buf, sizeof(error_buf),
		 "netstring %s (length %ld) incorrectly terminated: expected comma, got %c",
		 buf, *len_ptr, c);
	proto_error(error_buf);
    }
    
    return buf;
}


void verify_expected(FILE *stream, const char *expected)
{
    char *got;
    size_t len;
    char buff[256] = {0}; /* meant for error messages */
    
    got = read_netstring(stream, &len);
    if (strcmp(expected, got) != 0)
    {
	snprintf(buff, sizeof(buff), "expected %s, got %s",
		 expected, got);
	proto_error(buff);
    }
    
    free(got);
}


void usage(void)
{
    fprintf(stderr, "usage: %s [--stop] [args]\n", PROGNAME);
    exit(EX_USAGE);
}


void error(const char *msg)
{
    perror(msg);
    exit(EX_OSERR);
}


void proto_error(const char *msg)
{
    fprintf(stderr, "%s: %s: protocol error\n", PROGNAME, msg);
    exit(EX_PROTOCOL);
}


void send_fd(int s, int fd)
{
    struct msghdr   msg = {0};
    struct cmsghdr *cmsg;
    struct iovec    iov[1];
    
    char buf[CMSG_SPACE(sizeof(fd))] = {0};  /* ancillary data buffer */
    char identifier[] = {'X'};   /* baggage to help blocking kick in? */
    
    msg.msg_control    = buf;
    msg.msg_controllen = sizeof(buf);
    
    /* trying to send some bytes with it to help blocking kick in? */
    msg.msg_iovlen  = 1;
    msg.msg_iov     = iov;
    iov[0].iov_len  = sizeof(identifier);
    iov[0].iov_base = identifier;
    
    cmsg             = CMSG_FIRSTHDR(&msg);
    cmsg->cmsg_len   = CMSG_LEN(sizeof(fd));
    cmsg->cmsg_level = SOL_SOCKET;
    cmsg->cmsg_type  = SCM_RIGHTS;
    
    *(int *)CMSG_DATA(cmsg) = fd;
    
    if (sendmsg(s, &msg, 0) != sizeof(identifier))
    {
	error("sendmsg");
    }
}
